File: packages/python/src/sarj_python_lint/rules/single_public_export.py
"""SARJ022: rename a junk-drawer module stem with a single public export.

Such a module stem should describe its responsibility instead.

This rule is deliberately narrow. It fires ONLY when BOTH hold:

  (a) the module stem is a generic "junk-drawer" name that says nothing about
      the module's responsibility (`utils`, `base`, `models`, `types`, ...), AND
  (b) the module has exactly one public top-level `def` / `class`, so the
      rename target is unambiguous.

When both hold, the sole public export's name is the obvious, information-rich
replacement for the meaningless stem (`utils.py` exposing `snake_case_text` ->
`snake_case_text.py`; `enums.py` exposing `IntegrationProvider` ->
`integration_provider.py`).

Why the denylist gate matters: an informative stem names the module's DOMAIN
(`pagination.py`, `retry_wrapper.py`, `warmup.py`), which is frequently broader
than its one current export. Renaming those to the export loses the domain and
is a regression. A junk-drawer stem carries no domain to lose, so replacing it
with the export name is strictly an improvement.

Public = a top-level `class` / `def` / `async def` whose name has no leading
underscore. Constants (assignments), `__all__`, and imports are ignored.

Skipped entirely: `__init__.py`, `conftest.py`, test files (`test_*.py` or under
a `tests/` directory), and framework-convention filenames whose stem is fixed by
a framework/tool and cannot be renamed (`models.py`, `views.py`, `base.py`, ...).
Modules whose single export already snake-cases to the stem are not flagged
(there is nothing to improve).
"""

from __future__ import annotations

import ast
import re
from typing import TYPE_CHECKING, override

from sarj_python_lint.rule_base import Diagnostic, Rule, parse_or_none


if TYPE_CHECKING:
    from pathlib import Path


_SKIPPED_FILENAMES = frozenset({"__init__.py", "conftest.py"})

# Filenames whose stem is fixed by a framework or tool convention and therefore
# cannot be renamed without breaking discovery: Django reads models/views/urls/
# admin/apps/forms/settings/middleware/signals by filename, DRF `serializers.py`,
# Channels `routing.py`, Celery `tasks.py`, pytest `conftest.py`, `__main__.py`.
# Even when the stem is also a junk-drawer name (`models.py`, `base.py`), the
# rename the rule would suggest is not actionable, so these are never flagged.
_FRAMEWORK_CONVENTION_FILENAMES = frozenset(
    {
        "models.py",
        "admin.py",
        "apps.py",
        "views.py",
        "urls.py",
        "forms.py",
        "serializers.py",
        "base.py",
        "settings.py",
        "conftest.py",
        "__main__.py",
        "__init__.py",
        "middleware.py",
        "tasks.py",
        "signals.py",
        "routing.py",
    }
)

# Generic module stems that describe no responsibility. Curated conservatively:
# every entry is a name that, standing alone as a module, tells a reader nothing
# about what lives inside. Idiomatic domain stems (pagination, retry, warmup,
# client, service, ...) are deliberately excluded.
_JUNK_DRAWER_STEMS = frozenset(
    {
        "base",
        "common",
        "constant",
        "constants",
        "core",
        "enum",
        "enums",
        "helper",
        "helpers",
        "misc",
        "model",
        "models",
        "shared",
        "stuff",
        "type",
        "types",
        "util",
        "utils",
    }
)

# Multi-word acronyms whose community-accepted snake_case is a single token
# rather than the letter-by-letter split (`OAuth` -> `oauth`, not `o_auth`).
_ACRONYM_OVERRIDES: dict[str, str] = {"OAuth": "Oauth", "GraphQL": "Graphql", "gRPC": "Grpc"}

# Split on camelCase boundaries while keeping runs of capitals (acronyms)
# together: `HTTPServer` -> `HTTP` + `Server`, `JWTHandler` -> `JWT` + `Handler`.
_CAMEL_BOUNDARY_RE = re.compile(r"(?<=[a-z0-9])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])")


class SinglePublicExport(Rule):
    """Rename a junk-drawer module with a single public def/class after it."""

    id: str = "single-public-export"
    code: str = "SARJ022"
    description: str = "A junk-drawer module with a single public def/class should be renamed after that export."

    @override
    def check(self, path: Path, source: str) -> list[Diagnostic]:
        if _is_skipped_path(path):
            return []
        if path.stem.lower() not in _JUNK_DRAWER_STEMS:
            return []
        tree = parse_or_none(path, source)
        if tree is None:
            return []

        public_defs = [
            node
            for node in tree.body
            if isinstance(node, (ast.ClassDef, ast.FunctionDef, ast.AsyncFunctionDef)) and not node.name.startswith("_")
        ]
        if len(public_defs) != 1:
            return []

        primary = public_defs[0]
        expected_stem = _snake_case(primary.name)
        if path.stem == expected_stem:
            return []

        return [
            Diagnostic(
                path=path,
                line=primary.lineno,
                col=primary.col_offset + 1,
                code=self.code,
                message=(
                    f"module stem `{path.stem}` is a generic junk-drawer name; its sole public "
                    f"export is `{primary.name}` — rename the file to `{expected_stem}.py` to "
                    f"describe its responsibility."
                ),
            )
        ]


def _snake_case(name: str) -> str:
    for camel, replacement in _ACRONYM_OVERRIDES.items():
        name = name.replace(camel, replacement)
    return _CAMEL_BOUNDARY_RE.sub("_", name).lower()


def _is_skipped_path(path: Path) -> bool:
    if path.name in _SKIPPED_FILENAMES:
        return True
    if path.name.lower() in _FRAMEWORK_CONVENTION_FILENAMES:
        return True
    if path.name.startswith("test_"):
        return True
    return "tests" in path.parts

File: packages/typescript/src/rules/single-public-export.ts
/**
 * @fileoverview Flag a junk-drawer module stem that has a single public export.
 * TS port of Python SARJ022. Fires ONLY when BOTH hold:
 *
 *   (a) the file's basename stem is a generic "junk-drawer" name that describes
 *       no responsibility (`utils`, `helpers`, `types`, `models`, ...), AND
 *   (b) the module has exactly one public export, and that export is a named
 *       function / class / function-const, so the rename target is unambiguous.
 *
 * When both hold, the sole export's name is the information-rich replacement for
 * the meaningless stem (`utils.ts` exporting `snakeCaseText` -> `snake-case-text.ts`).
 *
 * A junk-drawer stem carries no domain to lose, so replacing it with the export
 * name is strictly an improvement; an informative stem (`pagination.ts`) names a
 * domain broader than its one current export and is deliberately never flagged.
 */

import { ESLintUtils, type TSESTree, AST_NODE_TYPES } from "@typescript-eslint/utils";

type MessageIds = "renameJunkDrawer";
type Options = readonly [];

// Generic module stems that describe no responsibility. `index` is deliberately
// excluded: barrel files legitimately re-export many symbols under that name.
const JUNK_DRAWER_STEMS = new Set([
  "util",
  "utils",
  "helper",
  "helpers",
  "common",
  "constant",
  "constants",
  "type",
  "types",
  "model",
  "models",
  "shared",
  "misc",
]);

// Exports whose name is an idiomatic ecosystem convention that lives in a
// junk-drawer bucket by design — flagging them fights the convention and the
// bucket is expected to grow. `cn` is the shadcn/ui tailwind-merge className
// helper that scaffolds into `lib/utils.ts`; renaming to `cn.ts` breaks every
// `import { cn } from "@/lib/utils"`.
const CONVENTIONAL_BUCKET_EXPORTS = new Set(["cn"]);

// Multi-word acronyms whose accepted kebab-case is a single token rather than a
// letter-by-letter split (`OAuth` -> `oauth`, not `o-auth`).
const ACRONYM_OVERRIDES: ReadonlyArray<readonly [RegExp, string]> = [
  [/OAuth/g, "Oauth"],
  [/GraphQL/g, "Graphql"],
  [/gRPC/g, "Grpc"],
];

// Split on camelCase boundaries while keeping runs of capitals (acronyms)
// together: `HTTPServer` -> `HTTP` + `Server`, `JWTHandler` -> `JWT` + `Handler`.
const CAMEL_BOUNDARY_RE = /(?<=[a-z0-9])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])/g;

const TEST_FILE_RE = /\.(test|spec)\.[cm]?[jt]sx?$/i;
const SCRIPT_EXT_RE = /\.[cm]?[jt]sx?$/i;

interface ExportSummary {
  readonly names: number;
  readonly hasReExport: boolean;
  readonly candidate: { name: string; node: TSESTree.Node } | null;
}

const basename = (filename: string): string =>
  filename.split(/[/\\]/).pop() ?? filename;

const stemOf = (base: string): string => base.replace(SCRIPT_EXT_RE, "");

const kebabCase = (name: string): string => {
  let normalized = name;
  for (const [pattern, replacement] of ACRONYM_OVERRIDES) {
    normalized = normalized.replace(pattern, replacement);
  }
  return normalized.replace(CAMEL_BOUNDARY_RE, "-").toLowerCase();
};

const isFunctionExpression = (node: TSESTree.Expression | null): boolean =>
  node !== null &&
  (node.type === AST_NODE_TYPES.ArrowFunctionExpression ||
    node.type === AST_NODE_TYPES.FunctionExpression);

// A `const foo = () => {}` is a rename candidate; a value const (`const MAX = 5`)
// is not — renaming a file after a bare constant loses more than it gains.
const functionConstName = (
  decl: TSESTree.VariableDeclaration,
): string | null => {
  if (decl.declarations.length !== 1) return null;
  const [declarator] = decl.declarations;
  if (declarator === undefined) return null;
  if (declarator.id.type !== AST_NODE_TYPES.Identifier) return null;
  if (!isFunctionExpression(declarator.init)) return null;
  return declarator.id.name;
};

const summarizeExports = (body: readonly TSESTree.ProgramStatement[]): ExportSummary => {
  let names = 0;
  let hasReExport = false;
  let candidate: { name: string; node: TSESTree.Node } | null = null;

  const addCandidate = (name: string, node: TSESTree.Node): void => {
    names += 1;
    candidate = { name, node };
  };

  for (const statement of body) {
    switch (statement.type) {
      case AST_NODE_TYPES.ExportAllDeclaration:
        hasReExport = true;
        break;
      case AST_NODE_TYPES.ExportDefaultDeclaration: {
        names += 1;
        const decl = statement.declaration;
        if (
          decl.type === AST_NODE_TYPES.FunctionDeclaration &&
          decl.id !== null
        ) {
          candidate = { name: decl.id.name, node: statement };
        } else if (
          decl.type === AST_NODE_TYPES.ClassDeclaration &&
          decl.id !== null
        ) {
          candidate = { name: decl.id.name, node: statement };
        }
        break;
      }
      case AST_NODE_TYPES.ExportNamedDeclaration: {
        if (statement.source !== null) {
          hasReExport = true;
          break;
        }
        const decl = statement.declaration;
        if (decl === null) {
          names += statement.specifiers.length;
          break;
        }
        switch (decl.type) {
          case AST_NODE_TYPES.FunctionDeclaration:
            if (decl.id !== null) addCandidate(decl.id.name, statement);
            else names += 1;
            break;
          case AST_NODE_TYPES.ClassDeclaration:
            if (decl.id !== null) addCandidate(decl.id.name, statement);
            else names += 1;
            break;
          case AST_NODE_TYPES.VariableDeclaration: {
            const fnName = functionConstName(decl);
            if (fnName !== null && decl.declarations.length === 1) {
              addCandidate(fnName, statement);
            } else {
              names += decl.declarations.length;
            }
            break;
          }
          default:
            names += 1;
        }
        break;
      }
      default:
        break;
    }
  }

  return { names, hasReExport, candidate };
};

export default ESLintUtils.RuleCreator(
  (name) =>
    `https://github.com/sarj-ai/standards/blob/main/packages/typescript/src/rules/${name}.ts`,
)<Options, MessageIds>({
  name: "single-public-export",
  meta: {
    type: "suggestion",
    docs: {
      description:
        "A junk-drawer module stem (`utils`, `helpers`, `types`, ...) with a single public function/class/const export should be renamed after that export.",
    },
    schema: [],
    messages: {
      renameJunkDrawer:
        "Module stem `{{stem}}` is a generic junk-drawer name; its sole public export is `{{name}}` — rename the file to `{{expected}}.ts` to describe its responsibility.",
    },
  },
  defaultOptions: [],
  create(context) {
    const base = basename(context.filename);

    if (base.endsWith(".d.ts")) return {};
    if (TEST_FILE_RE.test(base)) return {};

    const stem = stemOf(base);
    if (!JUNK_DRAWER_STEMS.has(stem.toLowerCase())) return {};

    return {
      Program(node: TSESTree.Program): void {
        const { names, hasReExport, candidate } = summarizeExports(node.body);
        if (hasReExport) return;
        if (names !== 1 || candidate === null) return;
        if (CONVENTIONAL_BUCKET_EXPORTS.has(candidate.name)) return;

        const expected = kebabCase(candidate.name);
        if (stem === expected) return;

        context.report({
          node: candidate.node,
          messageId: "renameJunkDrawer",
          data: { stem, name: candidate.name, expected },
        });
      },
    };
  },
});
